A55-xds-stateful-session-affinity.md

```diff
@@ -0,0 +1,516 @@
+a cookie in the RPC. gRPC routes that RPC to some backend B1 based on the
+configured LB policy. When gRPC receives the response for that first RPC from
+B1, it attaches a configured cookie to the response before delivering the
+response to the client application. The cookie encodes some identifying
+information about B1 (for example its IP-address). The client application
+saves the cookie and uses that cookie in all future RPCs of that session
+(e.g. by using a separate cookie jar for each session). gRPC uses the
+cookie and gets the encoded backend information to route the RPC to the
+same backend B1 if it is still reachable.
+
+With stateful session affinity, session draining of backends is an important
+requirement. In this use-case, the user wants to stop or restart service
+backends (let's say to optimize resources or to perform upgrades). However a
+terminating backend first needs to "drain" all assigned sessions before
+exiting. In this state - let's call it `DRAINING` - the terminating backend is
+only sent RPCs of its existing assigned sessions and no new traffic is sent.
+Once all assigned sessions have finished (or a timeout triggers), the backend
+application exits at which point the backend is now removed from the endpoints
+list (EDS). This design includes support for session draining.
+
+### Related Proposals: 
+
+* [A27: xDS-Based Global Load Balancing][A27]
+* [A42: xDS Ring Hash LB Policy][A42]
+
+## Proposal
+
+The goal of this design is to make it consistent with the Envoy implementation
+so it will interoperate with it. For example, one can create an xDS
+configuration for Envoy and the exact same configuration will work with
+proxyless gRPC. Similarly, an application using some cookie implementation
+that works with Envoy will continue to work with proxyless gRPC without any
+modifications.
+
+The design involves the following parts.
+
+### `XdsClient` to Include Endpoint Health Status
+
+The `EdsUpdate` struct from the `XdsClient` will be enhanced to include the
+[endpoint health status][eds-health-status] for each endpoint. `XdsClient`
+will also be modified to include endpoints in `DRAINING` state in the
+`EdsUpdate` reported to the watchers.
+
+### Listener and Route Configuration through xDS
+
+The feature is enabled and configured using the
+[cookie based stateful session extension][cookie-ext] which
+is an [http-filter][] in the `HttpConnectionManager` for the target service.
+A sample configuration is as follows:
+
+```textproto
+{ // HttpConnectionManager
+  http_filters: [
+    {
+      name: "envoy.filters.http.stateful_session"
+      typed_config: {
+        "@type": "type.googleapis.com/envoy.extensions.filters.http.stateful_session.v3.StatefulSession"
+        session_state: {
+          name: "envoy.http.stateful_session.cookie"
+          typed_config: {
+            "@type": "type.googleapis.com/envoy.extensions.http.stateful_session.cookie.v3.CookieBasedSessionState"
```

Comment:
What should the filter do if the value is not type.googleapis.com/envoy.extensions.http.stateful_session.cookie.v3.CookieBasedSessionState?

Reply:
Added